test/test_ur_urdf_xacro.py

```diff
@@ -0,0 +1,111 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+# Author: Lukas Sackewitz
+
+import os
+import shutil
+
+from ament_index_python.packages import get_package_share_directory
+
+
+def test_ur_urdf_xacro():
+
+    # Initialize Arguments
+    ur_type = "ur3"
+    safety_limits = "true"
+    safety_pos_margin = "0.15"
+    safety_k_position = "20"
+    # General Arguments
+    description_package = "ur_description"
+    description_file = "ur.urdf.xacro"
+    prefix = '""'
+
+    joint_limit_params = os.path.join(
+        get_package_share_directory(description_package), "config", ur_type, "joint_limits.yaml"
+    )
+    kinematics_params = os.path.join(
+        get_package_share_directory(description_package),
+        "config",
+        ur_type,
+        "default_kinematics.yaml",
+    )
+    physical_params = os.path.join(
+        get_package_share_directory(description_package),
+        "config",
+        ur_type,
+        "physical_parameters.yaml",
+    )
+    visual_params = os.path.join(
+        get_package_share_directory(description_package),
+        "config",
+        ur_type,
+        "visual_parameters.yaml",
+    )
+
+    description_file_path = os.path.join(
+        get_package_share_directory(description_package), "urdf", description_file
+    )
+
+    urdf_tmp_output = os.path.join(
+        get_package_share_directory(description_package), "test", "tmp_ur.urdf"
+    )
+
+    # Compose `xacro` command
+    xacro_command = (
+        f"{shutil.which('xacro')}"
+        f" {description_file_path}"
+        f" joint_limit_params:={joint_limit_params}"
+        f" kinematics_params:={kinematics_params}"
+        f" physical_params:={physical_params}"
+        f" visual_params:={visual_params}"
+        f" safety_limits:={safety_limits}"
+        f" safety_pos_margin:={safety_pos_margin}"
+        f" safety_k_position:={safety_k_position}"
+        f" name:={ur_type}"
+        f" prefix:={prefix}"
+        f" > {urdf_tmp_output}"
+    )
+    os.popen(xacro_command).read()
+
+    # Compose `check_urdf` command
+    check_urdf_command = f"{shutil.which('check_urdf')} {urdf_tmp_output}"
+    check_urdf_output = os.popen(check_urdf_command).read()
+
+    assert (
+        check_urdf_output != ""
```

Comment:
It would be better to check the executable's return code instead of matching the output string. That might be changing in the future and is therefore a very fragile test.

Reply:
I changed the subprocess calls to rely on [`subprocess`](https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module) and its returncode with commit ea29823f16f3a4656641003b051ddda9906df43f